File: packages/dart/lib/Seguranca/Ataques/AtaqueChaveErrada.dart
import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/CriptografiaAlvo.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/SkipAtaqueException.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// Garante que um token só decifra com a chave correta: com qualquer outra
/// chave de 32 bytes, decrypt() tem que lançar. Verifica também que a chave
/// errada não "quase funciona" (ex: aceitar às vezes por MAC fraco).
///
/// Cuidado: CriptografiaAlvo escreve a chave no estado global; a chave
/// original é restaurada no final pra não afetar os outros ataques.
class AtaqueChaveErrada extends Ataque {
  final int _tentativas;

  AtaqueChaveErrada({this._tentativas = 30});

  @override
  String nome() => 'Rejeição de chave incorreta';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    if (alvo is! CriptografiaAlvo) {
      throw SkipAtaqueException(
        'Precisa de CriptografiaAlvo para trocar a chave.',
      );
    }

    final chaveOriginal = alvo.chaveDeTeste();

    final tokens = <String>[];
    for (var i = 0; i < _tentativas; i++) {
      tokens.add(alvo.encrypt('MENSAGEM_SECRETA_$i'));
    }

    final aceitos = <String>[];
    try {
      for (var i = 0; i < tokens.length; i++) {
        CriptografiaAlvo(bytesToHex(randomBytes(16)));
        try {
          final r = alvo.decrypt(tokens[i]);
          aceitos.add('tentativa $i aceitou (retornou $r)');
        } catch (_) {
          // esperado
        }
      }
    } finally {
      CriptografiaAlvo.definirChaveGlobal(chaveOriginal);
    }

    if (aceitos.isNotEmpty) {
      return ResultadoAtaque(
        nome(),
        true,
        Severidade.critica,
        '${aceitos.length} de $_tentativas tokens foram aceitos com a chave '
            'errada',
        {'exemplos': aceitos.take(5).toList()},
      );
    }

    return ResultadoAtaque(
      nome(),
      false,
      Severidade.info,
      'Nenhum dos $_tentativas tokens foi aceito com chave incorreta',
    );
  }
}

File: packages/rust/src/seguranca/ataques/ataque_chave_errada.rs
//! Garante que um token só decifra com a chave correta.

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::criptografia_alvo::CriptografiaAlvo;
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};
use crate::seguranca::util::{hex_encode, random_bytes};

pub struct AtaqueChaveErrada {
    pub tentativas: usize,
}

impl AtaqueChaveErrada {
    pub fn new() -> Self {
        AtaqueChaveErrada { tentativas: 30 }
    }
}

impl Default for AtaqueChaveErrada {
    fn default() -> Self {
        Self::new()
    }
}

impl Ataque for AtaqueChaveErrada {
    fn nome(&self) -> String {
        "Rejeição de chave incorreta".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        if alvo.as_any().downcast_ref::<CriptografiaAlvo>().is_none() {
            return Err(ErroAtaque::Skip(
                "Precisa de CriptografiaAlvo para trocar a chave.".to_string(),
            ));
        }

        let chave_original = alvo.chave_de_teste();

        let mut tokens: Vec<String> = Vec::new();
        for i in 0..self.tentativas {
            tokens.push(
                alvo.encrypt(format!("MENSAGEM_SECRETA_{}", i).as_bytes())
                    .map_err(|e| ErroAtaque::Falha(e.to_string()))?,
            );
        }

        let mut aceitos: Vec<String> = Vec::new();
        for (i, token) in tokens.iter().enumerate() {
            let chave_errada = hex_encode(&random_bytes(16));
            CriptografiaAlvo::com_chave(&chave_errada);
            if let Ok(r) = alvo.decrypt(token) {
                aceitos.push(format!("tentativa {} aceitou (retornou {})", i, r));
            }
        }
        CriptografiaAlvo::restaurar_chave(&chave_original);

        if !aceitos.is_empty() {
            return Ok(ResultadoAtaque::new(
                &self.nome(),
                true,
                Severidade::Critica,
                &format!(
                    "{} de {} tokens foram aceitos com a chave errada",
                    aceitos.len(),
                    self.tentativas
                ),
            ));
        }

        Ok(ResultadoAtaque::new(
            &self.nome(),
            false,
            Severidade::Info,
            &format!(
                "Nenhum dos {} tokens foi aceito com chave incorreta",
                self.tentativas
            ),
        ))
    }
}
